gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+                        continue
+                    ev = val['queue'][0]
+                    LOG.debug(_("Returing serialized event %s"
+                                % (ev.identify())))
+                    return ev
+        return None
+
+    def workers_init(self):
+        wc = 2 * (multiprocessing.cpu_count())
+        if cfg.CONF.workers != wc:
+            wc = cfg.CONF.workers
+            LOG.info(_("Creating %d number of workers" % (wc)))
+
+        workers = [tuple() for w in range(0, wc)]
+
+        for w in range(0, wc):
+            qu = Queue()
+            evworker = EventHandler(self, qu, self.ehs)
+            proc = Process(target=evworker.run, args=(qu,))
+            proc.daemon = True
+            workers[w] = workers[w] + (proc, qu, evworker)
+        return workers
+
+    def poll_init(self):
+        qu = Queue()
+        ph = PollHandler(self, qu, self.ehs)
+        return ph
+
+    def modules_init(self, modules):
+        for module in modules:
+            LOG.info(_("Initializing module %s" % (identify(module))))
+            try:
+                module.module_init(self, self._conf)
+            except AttributeError as s:
+                LOG.error(_("Module %s does not implement"
+                            "module_init() method - skipping"
+                            % (identify(module))))
+                continue
+                # raise AttributeError(module.__file__ + ': ' + str(s))
+        return modules
+
+    def _init(self):
+        self.ehs = EventHandlers()
+        self.rpc_agents = RpcAgents()
+        self.modules = self.modules_init(self.modules)
+        self.workers = self.workers_init()
+        self.pollhandler = self.poll_init()
+        self.loadbalancer = getattr(
+            globals()['core_lb'], cfg.CONF.RpcLoadBalancer)(self.workers)
+
+    def wait(self):
+        # self.rpc_agents.wait()
+        for w in self.workers:
+            w[0].join()
+
+    def start(self):
+        self._init()
+        self.periodic_task = PeriodicTask(self)
+        # Seperate task for reporting as report state rpc is a 'call'
+        self.reportstate_task = ReportStateTask(self)
+
+        for w in self.workers:
+            w[0].start()
+            LOG.debug(_("Started worker - %d" % (w[0].pid)))
+        self.rpc_agents.launch()
+
+    def rpc_event(self, event):
+        worker = self.loadbalancer.get(event.binding_key)
+        event.worker_attached = worker[0].pid
+        LOG.info(_("Scheduling internal event %s"
+                   "to worker %d"
+                   % (event.identify(), event.worker_attached)))
+        qu = worker[1]
+        qu.put(event)
+
+    def poll_event(self, event, max_times=sys.maxint):
+        LOG.info(_("Adding to pollq - event %s for maxtimes: %d"
+                   % (event.identify(), max_times)))
+        event.max_times = max_times
+        self.pollhandler.add(event)
+
+    def poll_event_done(self, event):
```

Comment:
Can this be the same as poll_event? Caller sets event.id = ..DONE and calls with max_times = 1?